### PR TITLE
Cut several dependencies of OrbitService on globals and client-only code

### PR DESCRIPTION
**Pull request by dpallotti**
#### Split SymbolsManager in SymbolsService,Client
These were two separate interfaces for the client and the server in the same
class.
Rename `CoreApp::InitializeManagers` to `InitializeClientTransactions` and use
it only in the ui, move the server-related logic to OrbitService.
#### Move GetRemoteMemory logic to new ProcessMemoryClient,Service
Replace `CoreApp::SetupClientMemoryTransaction`,
`OrbitService::SetupServiceMemoryTransaction` with new classes
`ProcessMemoryClient`, `ProcessMemoryService`.
#### Split TransactionManager in TransactionClient,TransactionService
Only use `TransactionClient` in the client, only `TransactionService` in
`OrbitService`.
#### Change OrbitService to OrbitAsioServer. Use the former to manage both GrpcServer and AsioServer
---
The overall goal of this change is to simplify `OrbitService`, removing dependencies on (or at lest not use) code and classes that should be client-specific. This should make it easier to understand the code flow in the service.
In particular, classes that have a double-interface for the client and the service are split in two. This also allows to reduce the number of globals that `OrbitService` uses (the only one left is `GTcpServer`).

### CODE REVIEW

File: OrbitService/OrbitService.cpp [hunk 1]
```diff
@@ -1,27 +1,165 @@
...
+      &process_list_, transaction_service_.get());
   process_memory_service_ =
       std::make_unique<ProcessMemoryService>(transaction_service_.get());
 }
 
-void OrbitService::Run(std::atomic<bool>* exit_requested) {
-  while (!(*exit_requested)) {
-    GTcpServer->ProcessMainThreadCallbacks();
-    Sleep(16);
+void OrbitService::TracingBufferThread() {
```
Comment (dimitry-) [inline OrbitService/OrbitService.cpp:121]: Most of this code should probably be moved to OrbitAsioService - since this is asio specific code.

**Comment by dpallotti** [comment]
> In the last commit, can you please move server specific code to OrbitAsioService instead of OrbitService?

@dimitry- Do you mean that `OrbitService` should be renamed to `OrbitAsioService` or that `OrbitService` should be split in two?

File: OrbitCore/CoreApp.h [hunk 2]
```diff
@@ -61,27 +62,21 @@ class CoreApp {
...
-  bool IsService() const { return is_service_; }
+  SymbolsClient* GetSymbolsClient() { return symbols_client_.get(); }
 
  private:
   // Transactions
-  void SetupMemoryTransaction();
-  typedef std::function<void(std::vector<byte>&)> memory_callback;
-  absl::flat_hash_map<uint32_t, memory_callback> memory_callbacks_;
+  void SetupClientMemoryTransaction();
+  typedef std::function<void(std::vector<byte>&)> MemoryCallback;
```
[the remark this answers is not in the document]
Comment (dpallotti) [reply OrbitCore/CoreApp.h]: This is changed by a following commit.

File: OrbitCore/CoreApp.h [hunk 3]
```diff
@@ -61,27 +62,21 @@ class CoreApp {
...
+  absl::flat_hash_map<uint32_t, MemoryCallback> memory_callbacks_;
   absl::Mutex transaction_mutex_;
 
   std::unique_ptr<orbit::TransactionManager> transaction_manager_ = nullptr;
-  std::unique_ptr<orbit::SymbolsManager> symbols_manager_ = nullptr;
-
-  bool is_client_;
-  bool is_service_;
+  std::unique_ptr<SymbolsClient> symbols_client_ = nullptr;
```
[the remark this answers is not in the document]
Comment (dpallotti) [reply OrbitCore/CoreApp.h]: Removed.

File: OrbitCore/CoreApp.cpp [hunk 4]
```diff
@@ -10,15 +10,14 @@
...
   transaction_manager_ = std::make_unique<orbit::TransactionManager>(
       GTcpClient.get(), GTcpServer.get());
-  symbols_manager_ = std::make_unique<orbit::SymbolsManager>(GCoreApp);
-
-  SetupMemoryTransaction();
-
-  is_client_ = ConnectionManager::Get().IsClient();
-  is_service_ = ConnectionManager::Get().IsService();
+  symbols_client_ =
+      std::make_unique<SymbolsClient>(GCoreApp, transaction_manager_.get());
```
[the remark this answers is not in the document]
Comment (dpallotti) [reply OrbitCore/CoreApp.cpp]: Yep

File: OrbitCore/SymbolsClient.h [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  void LoadSymbolsFromSession(Process* process,
+                              const std::shared_ptr<Session>& session);
+
+ private:
+  void HandleResponse(const Message& message, uint32_t id);
+
+  CoreApp* core_app_ = nullptr;
+  orbit::TransactionManager* transaction_manager_;
+  absl::flat_hash_map<uint32_t, std::shared_ptr<Session>> id_sessions_;
+  absl::Mutex mutex_;
```
[the remark this answers is not in the document]
Comment (dpallotti) [reply OrbitCore/SymbolsClient.h]: Agree

**Comment by dimitry-** [comment]
> > In the last commit, can you please move server specific code to OrbitAsioService instead of OrbitService?
> 
> @dimitry- Do you mean that `OrbitService` should be renamed to `OrbitAsioService` or that `OrbitService` should be split in two?

I was actually thinking about moving code from ConnectionManager to OrbitAsioService (newly introduced) and keep OrbitService as top level which for the time being runs Grpc and Asio services.

**Comment by dpallotti** [comment]
> > > In the last commit, can you please move server specific code to OrbitAsioService instead of OrbitService?
> > 
> > 
> > @dimitry- Do you mean that `OrbitService` should be renamed to `OrbitAsioService` or that `OrbitService` should be split in two?
> 
> I was actually thinking about moving code from ConnectionManager to OrbitAsioService (newly introduced) and keep OrbitService as top level which for the time being runs Grpc and Asio services.

Ah right, so the use of `OrbitGrpcServer` should be moved from `main` to `OrbitService`.

File: OrbitService/OrbitService.cpp (hunk 1 above)
Comment (dpallotti) [reply OrbitService/OrbitService.cpp]: I added a commit for that, please have a look.

File: OrbitCore/TransactionClient.h [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+//
+// Usage: Register one TransactionResponseHandler per message type on the
+// TransactionClient and one TransactionRequestHandler per message type on the
+// TransactionService. A TransactionRequestHandler carries a request handler
+// that will be executed on the service side and TransactionResponseHandler
+// carries a response handler for the client side.
+// The steps for issuing requests and receiving a response are:
+// 1. The client enqueues a request through "TransactionClient::EnqueueRequest".
+// 2. The TransactionClient schedules and sends out the request.
+// 3. Then service receives the request in its request handler.
```
[the remark this answers is not in the document]
Comment (dpallotti) [reply OrbitCore/TransactionClient.h]: Done.

**Pull request merged**